README.md
<!-- Title -->
<h1 align="center">
  SubZero
</h1>

<!-- description -->
<p align="center">
  <strong> Sea ice model to explicitly simulate individual floe life cycles using complex discrete elements with time-evolving shapes.</strong>
</p>
<!-- Information badges -->
<p align="center">
  <a href="https://www.repostatus.org/#active">
    <img alt="Repo status" src="https://www.repostatus.org/badges/latest/active.svg?style=flat-square" />
  <a href="https://doi.org/10.5281/zenodo.7199941">
    <img alt="Zenodo" src="https://zenodo.org/badge/doi/10.5281/zenodo.7199941.svg">
  <a href="https://github.com/bpm5026/SubZero_bpm/actions/workflows/ci.yml/badge.svg?branch=JOSS">
    <img alt="GitHub Actions CI Status" src="https://github.com/bpm5026/SubZero_bpm/actions/workflows/ci.yml/badge.svg?branch=JOSS">
  </a>
</p>

This unique model uses parameterizations of floe-scale processes, such as collisions, fractures, ridging, and welding, to by pass resolving intra-floe bonded elements. We demonstrate the novel capabilities of the SubZero model in idealized experiments, including uniaxial compression, the summer-time sea ice flow through the Nares Strait and a winter-time equilibration of floe size and
ice thickness distributions.

For more information see paper: A Sea Ice Model with an Explicit Representation of the Floe Life Cycle by Georgy Manucharyan and Brandon Montemuro. 
This code is also saved on Zenodo

## Installation instructions

1. Clone SubZero contents to their own folder. Open folder within MATLAB.

2. Once within the MATLAB project, open the MATLAB command window.

3. Run command ```cd private``` to enter `private` folder within the command window.

4. Run command ```mex('-D__int64=__int64_t','clipper.cpp','mexclipper.cpp')``` in MATLAB command window. This will link the C++ code used for polygon clipping within the program. For more information on this linking see [here](https://www.mathworks.com/help/matlab/ref/mex.html).

5. If this does not work, you may have to install Xcode to compile the mex file on a Mac. Xcode is a large file and is available in App Store and online. Downloading the installer from the apple website may be quicker.

6. After installing Xcode, you may have to run these two lines in the terminal: 
    + ```sudo xcode-select -s /Applications/Xcode.app/Contents/Developer```
    + ```sudo xcodebuild -license accept```
   
 7. Shut down MATLAB, re-open, and re-do step 4.
 
 ## Running your first model

Model runs out of the box. Currently set up model creates ~10 floes that fill the entire domain with collisions and corner-griding on, and all other functionality turned off. Model will run for 7500 timesteps where each timestep is 10 seconds.

To change the model, the following steps can be taken:

1. Define the boundaries of the domain `initialize_boundaries.m` file. Three options are present, with two commented out. Choose one of the given options, or write your own.

2. Set up desired ocean forcings in `initialize_ocean.m` file. Specifically see the `psi_ocean`, `Uocn`, and `Vocn` variables. Note that your ocean must be bigger than the boundaries of the domain, in some cases significantly bigger to prevent errors with periodic boundaries. Additionally, the user is responsible for making the ocean doubly periodic outside of domain bounds. 

3. Define the atmospheric winds within the `SubZero.m` file. Specifically see the `winds.u` and `winds.v` variables.

4. Define the concentration and number of floes you wish to start with in the SubZero.m file. The concentration is the fraction of the domain that will initally be covered in ice. For this, see the variable `target_concentration` around line 57. This variable is repurposed below for concentration when packing (re-set to 1), so the initial value must be set for model initialization. The number of floes is the fourth argument in the following line:
  `[Floe, Nb] = initial_concentration(c2_boundary,target_concentration,height,10,min_floe_size)` - here 10 floes.

5. Within `initial_concentration`, topography can be added to the simulation. These will be represented as immovable, unbreakable floes. `Nb` is the number of these boundaries and their shape can be input here as well. See Nares Strait example within the `validation_cases` folder to see how to add these elements.

5. Set the flags for desired physical processes to `true` as well as the frequency with which you want them to execute in `SubZero.m` file. The flags are at the beginning of the file, while the frequencies are lower throughout the file. Look for lines of the form `if FRACTURES && mod(i_step,75)==0` with different flags. These lines give the frequency of each of these flag operations. In the example above, fractures happen every 75 timesteps.

7. Set the timestep `dt` directly under the flags.

6. Set the total number of snapshots of the system as well as how often that will be output in `SubZero.m` file. These are set under the `Set output parameters` comment.

7. Run the `SubZero.m` program to start a simulation. 

To see examples of different scenarios, see the examples included in `validation_cases` folder as tar files. These are the validation cases from the Subzero paper.  

## Validation Cases  
### 1. Uniaxial Compression:
Here we demonstrate the behavior of sea ice floes subject to uniaxial compression in a confined domain. The run is initialized with 200 floes in a fully-packed domain with the North/South boundaries moving towards the center of the domain, and stationary East/West boundaries. A relatively small time step, dt=5 s, is used to resolve the elastic waves in response to external boundary motion and changes in the floe configuration due to fractures. The atmospheric and oceanic stresses are set to zero for this simplified test. The floes are subject to Mohr-Coulomb fracture criteria, but there is no floe simplification, corner grinding, welding, ridging, rafting, or creation of new floes in this scenario. The boundaries move with a constant prescribed velocity, v_b = 0.1 m/s. To modify the base program to run the uniaxial compression validation case on your own, the following changes must be made.

1. In ```Subzero.m``` set:
  
    1a. Flags to false except FRACTURES and COLLISION to true. 

    1b. Set dt = 5;

    1c. The variables ```height.mean``` is set to 1 and `height.delta' should be set to 0. 

    1d. Atospheric winds should be U0 = 0 and V0 = 0. 

    1e. Target_concentration = 1

    1f. The number of initial floes to 200

    1g. The modulus = 2.5e3*(mean(sqrt(cat(1,Floe).area))+min(sqrt(cat(1,Floe.area)))

    1h. Set all instances of doInt.flag = false so there are no forces from ocean/atmosphere and all movement is driven by boundaries.

    1i. The frequency for fractures is set to mod(i_step,200)

    1j. After updating ```Time=Time+dt; i_step=i_step+1;``` add the following lines to have the boundary move.

        if max(c2_boundary(2,:)) > 85000 && mod(i_step,30)==0
            xb = c2_boundary(1,:);  
            yb = c2_boundary(2,:);  
            yb = yb - 15*[-1 1 1 -1]; 
            c2_boundary = [xb; yb];  
            Ly = max(c2_boundary(2,:));Lx = max(c2_boundary(1,:));  
            c2_boundary_poly = polyshape(c2_boundary');  
            c2_border = scale(c2_boundary_poly,2); c2_border = subtract(c2_border, c2_boundary_poly);
            floebound = initialize_floe_values(c2_border, height);  
        end
  
2. In ```initialize_boundaries.m``` the box is set such that Lx=Ly=100 kilometers.
  
3. In ```fracture.m``` set Sig11 = 1.5e5 and make sure all the elliptical yield curve lines are commented out or deleted.  

4. In ```floe_interactions.m``` set mu = 0.3.
  
5. To do the other cases, rerun with modulus using 2.0e3 and 1.5e3 in place of 2.5e3.
  
### 2. Nares Strait:
The Nares Strait simulation demonstrates the role of floe fractures in wind-driven sea ice transport through narrow straits. Nares Strait is a channel between Ellesmere Island (Canada) and Greenland. The simulation aims to reflect spring or summer-like conditions of Arctic sea ice export through Nares Strait after the breakup of its winter arches. Since the transport events are relatively short (order of days or less), the effects of thermodynamic sea ice melt could be considered secondary relative to mechanical floe processes such as collisions and fractures. We thus randomly initialize the model with relatively large floes of uniform thickness, covering only the area just north of the strait. The uniform 10 m/s southward winds generate stresses that push the floes through the strait, while the ocean is assumed to be stagnant. Coastal boundaries are prescribed using a series of static floes. All physical processes except collisions and fractures are turned off to model the spring/summer breakup of floes. To modify the base program to run the Nares validation case on your own, the following changes must be made.
  
1. In ```Subzero.m``` set
    
      1a. flags to false except FRACTURES and COLLISION to true
      
      1b. The variables ```height.mean``` is set to 1 and ```height.delta``` should be set to 0
  
      1c. Atospheric winds should be U0 = 0 and V0 = -10.
  
      1d. Target_concentration = [1; 0]
  
      1e. The number of initial floes to 150 
  
      1f. The frequency for fractures is set to mod(i_step,150)

2. In ```initialize_ocean.m``` the ocean grid is set such that dXo=20000 meters, Lx=Ly =2e6 meters, and ocean velocity field is set to zero.
  
3. In ```initialize_boundaries.m``` the box is set such that Lx=50 kilometers, the top y boundary is 500 kilometers and the bottom y boundary is -250 kilometers.
  
4. In ```initialize_concentration.m``` 
  
    4a. Add the following lines after ```dy = y(2) -y(1)``` to create the boundaries of Nares Strait: (Note that you will need to include the file `Nares_Strait_segments.mat' in a folder named Nares. This mat file can be found in the Nares validation zip file)
  
        Lx = abs(min(x));Ly = abs(min(y));
        Bx = [-Lx -Lx Lx    Lx];  
        By = [-Ly  -1.5e5 -1.5e5  -Ly];
        B = polyshape(Bx',By');  

        %Create floes that act as boundaries and wont move
        load( './Nares/Nares_Strait_segments','R')

        %Remove islands
        if ~ISLANDS
          R(1:4) = [];
        end

        Floe = []; bound = c2_boundary_poly;

        for ii = 1:length(R)
          FloeNEW = initialize_floe_values(R(ii),height);    
          bound = subtract(bound, FloeNEW.poly);
          Floe = [Floe FloeNEW];
        end

        N = length(Floe);
        for ii = 1:N  
          Floe(ii).poly = polyshape(Floe(ii).c_alpha'+[Floe(ii).Xi Floe(ii).Yi]);  
        end
        Bu = union([Floe.poly]);
        Bu = union(Bu,B);

    4b. Then, after ```[~, b,~,~,~] = polybnd_voronoi([X Y],boundary.Vertices)``` insert  the following lines to remove any overlap between the shapes generated by the Voronoi tesselation and Nares Strait topography.

         for kk = 1:length(b)
           poly(kk) = polyshape(b{kk});
         end  
         poly = intersect([poly],polyB);
         polyout=subtract([poly],Bu);
         clear polyIce; polyIce = [];
         for kk = 1:length(polyout)
            R = regions(polyout(kk));
            polyIce = [polyIce; R];
         end
  
    4c. Replace the line ```Nf = 1:length(b);``` with the following:
  
          Nf = 1:length(polyIce);%randperm(length(b));
  
    4d. Replace the while loop with the following since we have already created the polyshapes;
  
        while Atot/area(polyB)<=c(jj,ii)                              
          floenew = initialize_floe_values(polyIce(Nf(count)),height);
          Floe = [Floe floenew];
          count = count+1;                                  
          Atot = Atot+floenew.area;        
          if count > length(Nf)  
              Atot = area(polyB)+1;
          end  
        end
 
  
    4e. Finally, after the lines ```areas = cat(1,Floe.area)``` add ```areas(1:Nb) = min_floe_size;```

5. In ```floe_interactions_all.m``` update to include c2_boundary such that it reads 
  [tmp,Fx,Fy] =calc_trajectory(dt,ocean,winds,Floe(i),HFo,doInt,c2_boundary);
  
6. In ```calc_trajectory.m```:
  
      6a. The first line needs changed so the function can take in the addtional input and should read:
  
          function [floe,FxOA,FyOA] =calc_trajectory(dt,ocean,winds,floe,HFo, doInt,c2_boundary)
  
      6b. After calculating the new floe.c_alpha add the following if statement:
  
          if min(floe.c_alpha(2,:))+floe.Yi<min(c2_boundary(2,:))
            floe.alive = 0;
          end
                                                                
 7. In ```fracture.m``` set Pstar = 1e5 and make sure all the Mohr's cone lines are commented out or deleted.                                                    
 8. In ```floe_interactions.m``` set mu = 0.25. 
                                                                
### 3. Winter ITD and FSD equilibration:
Here we demonstrate an essential case of model equilibration in winter-like conditions, where all parameterizations are active. We subject sea ice to strong mechanical and thermodynamic forcing over a five week period to facilitate an accelerated model evolution away from the initialized floe shapes, sizes, and thicknesses towards typical winter-like distributions. Specifically, we prescribe idealized ice-ocean stresses in the form of four equal-strength counter-rotating gyres (arranged like mechanical gears) that create relative sea ice motion and facilitate floe fractures and ridging. Alternatively, one could prescribe atmosphere-ocean stresses to achieve the same goal, but in this run the winds are set to 0. To make this a winter-like simulation, we ensured continuous sea ice growth by specifying a fixed negative heat flux that increases the thickness of existing ice floes, the formation of new ice floes in open ocean regions, and welding between floes.To modify the base program to run the winter validation case on your own, the following changes must be made.

 1. In ```Subzero.m``` set
      1a. RIDGING, FRACTURES, PERIODIC, PACKING, WELDING, CORNERS, COLLISION, RAFTING, and KEEP_MIN flags to true
      
      1b. The variables ```height.mean``` is set to 0.25 and ```height.delta``` should be set to 0
            
      1c. nDTpack = 5500
  
      1d. Atospheric winds should be U0 = 0 and V0 = 0.
  
      1e. Target_concentration = 1
  
      1f. The number of initial floes to 100 
            
      1g. nSimp = 20
           
 2. In ```initialize_boundaries.m``` the box is set such that Lx=Ly=100 kilometers
            
 3. In ```fractures.m``` set Pstar = 2.25e5 and make sure all the Mohr's cone lines are commented out or deleted. 

 4. In ```floe_interactions.m``` set mu = 0.3.
                                                                  
## Community Guidelines

If you experience issues using Subzero, please open a GitHub issue. If you're interested in contributing, please reach out to the paper authors. 


CHANGES.md
1.0.3 (April 3, 2023)
==================

- Updates for JOSS paper, including updates to README.md, fixing a bug in calculating collisions between ghost floes, the addition of an automated test, and additional documentation. New documentation includes the CHANGES.md file and CONTRIBUTING.md.


1.0.2 (Nov 18, 2022)
==================

- Update README.md


1.0.1 (Oct 18, 2022)
==================

- Updated SubZero release with ocean turn angle.


1.0.0 (Oct 14, 2022)
==================

- The initial release of the SubZero discrete element sea ice model.


Subzero.m
close all; clear all;

paths
%% Set Flags

RIDGING=false; 

FRACTURES=false;

PERIODIC=false;

PACKING = false;

WELDING = false;

CORNERS = true;

COLLISION = true;

AVERAGE = false; %average certain values in time

RAFTING = false;

KEEP_MIN = false; %Retain floes in simulation that get smaller set threshold

ifPlot = true; %Plot floe figures or not?

ifPlotStress = false; %Plot floe figures with floes shaded by stress values

ifPlotStressStrain = false; %Create plot with homogenized stress and strain values

justfrac = false;

%% Initialize model vars

dt=10; %Time step in sec
height.mean = 0.25; %mean value of thickness for initial floes
height.delta = 0; %maximum deviation about the mean thickness if a distribution is desired


nDTpack = 500; %Time streps between creation of new floes
rho_ice=920; % kg/m3
[ocean, HFo, h0]=initialize_ocean(dt,nDTpack); %Define ocean currents

%Define Atmospheric winds
winds.x = ocean.Xocn; winds.y = ocean.Yocn;
U0 = 0; V0 = 0;
winds.u=U0*ones(size(ocean.Xocn));
winds.v=V0*ones(size(ocean.Xocn));

%Define boundaries
c2_boundary=initialize_boundaries();
f=2*pi/(24*3600);
Ly = max(c2_boundary(2,:));Lx = max(c2_boundary(1,:));
min_floe_size = 4*Lx*Ly/10000; %set minimum floe size for initialization

%Initialize Floe state
target_concentration = 1; %Set target concentration for initial conditions
[Floe, Nb] = initial_concentration(c2_boundary,target_concentration,height,10,min_floe_size);

%create polyshape of any boundary floes
if Nb >0
    polyboundary = union([Floe(1:Nb).poly]); 
else
    polyboundary = [];
end
c2_boundary_poly = polyshape(c2_boundary');
c2_border = polyshape(2*[-Lx -Lx Lx Lx; -Ly Ly Ly -Ly]'); c2_border = subtract(c2_border, c2_boundary_poly);
floebound = initialize_floe_values(c2_border, height);
if isfield(Floe,'poly')
    Floe=rmfield(Floe,{'poly'});
end
min_floe_size = (4*Lx*Ly-sum(cat(1,Floe(1:Nb).area)))/20000; %set minimum floe size for simulation

%calculate elastic modulus used in floe interactions
global Modulus
Modulus = 1.5e3*(mean(sqrt(cat(1,Floe.area)))+min(sqrt(cat(1,Floe.area))));
save('Modulus.mat','Modulus');

%% Set output parameters 

dhdt = 1; %atmospheric heat flux

nDTOut=150; %Output frequency (in number of time steps)

nSnapshots=50; %Total number of model snapshots to save

nDT=nDTOut*nSnapshots; %Total number of time steps

nSimp = 20; %Timesteps between simplification of floe boundaries

% Set parallel computing inputs
%nPar = 18; %Number of workers for parfor
%poolobj = gcp('nocreate'); % If no pool, do not create new one.
%if isempty(poolobj)
%    parpool(nPar);
%else
%    delete(poolobj);
%    parpool(nPar);
%end

target_concentration=1; %Set target concentration for when creation of new elements will be run
tStart = tic; 

%ocean forces need updated on less frequent time scale
doInt.flag = true; 
doInt.step = 10; %Timesteps between update of ocean stresses

% specify coarse grid size
LxO= 2*max(ocean.Xo);LyO= 2*max(ocean.Yo);
Nx=10; Ny=10;%fix(Nx*LyO/LxO);
xc = min(c2_boundary(1,:)):(max(c2_boundary(1,:))-min(c2_boundary(1,:)))/Nx:max(c2_boundary(1,:));
yc = min(c2_boundary(2,:)):(max(c2_boundary(2,:))-min(c2_boundary(2,:)))/Ny:max(c2_boundary(2,:));
Xc = (xc(1:end-1)+xc(2:end))/2; Yc = -(yc(1:end-1)+yc(2:end))/2;

%initialize dissolved ice at zero
dissolvedNEW=zeros(Ny,Nx);

%Initiailize Eulearian Data
[eulerian_data] = calc_eulerian_data(Floe,Nx,Ny,Nb,c2_boundary,PERIODIC);
Vd = zeros(Ny,Nx,2);
Vdnew=zeros(Ny, Nx);
SigXX = zeros(Ny, Nx); SigYX = zeros(Ny, Nx);
SigXY = zeros(Ny, Nx); SigYY = zeros(Ny, Nx);
Eux = zeros(Ny, Nx); Evx = zeros(Ny, Nx);
Euy = zeros(Ny, Nx); Evy = zeros(Ny, Nx);
U = zeros(Ny, Nx); V = zeros(Ny, Nx);
dU = zeros(Ny, Nx); dV = zeros(Ny, Nx);
Sig = zeros(Ny, Nx); mass = zeros(Ny,Nx);

%% Calc interactions and plot initial state
Floe=Floe(logical(cat(1,Floe.alive)));
[Floe,dissolvedNEW] = floe_interactions_all(Floe, floebound, ocean, winds,c2_boundary, dt,HFo,min_floe_size,Nx,Ny,Nb, dissolvedNEW,doInt,COLLISION, PERIODIC, RIDGING, RAFTING); % find interaction points
A=cat(1,Floe.area);
Amax = max(A);

%% Initialize time and other stuff to zero
if isempty(dir('figs')); disp('Creating folder: figs'); mkdir('figs'); end
if isempty(dir('Floes')); disp('Creating folder: Floes'); mkdir('Floes'); end

if ~exist('Time','var')
    Time=0;
    i_step=0;
    im_num=1;
    fig=0;
    fig2=figure('Position',[100 100 1000 500],'visible','on');
end

%% Solving for floe trajectories
tic; 
while im_num<nSnapshots

    if mod(i_step,10)==0
        disp(' ');
        toc
        disp([num2str(i_step) ' timesteps comleted']); 
        numCollisions = calc_collisionNum(Floe);
        sacked = sum(~cat(1, Floe.alive));
        if sacked>0, disp(['sacked floes: ' num2str(sacked)]); end
        disp(['number of collisions: ' num2str(numCollisions)]);
        disp(' ');
        tic
        doInt.flag=true;
    else
        doInt.flag=false;
    end

    %Dynamic simplification of floe shape
    if mod(i_step,nSimp)==0
        FloeOld = Floe;
        parfor j=1:length(FloeOld)
            FloeOld(j).poly = polyshape(FloeOld(j).c_alpha'+[FloeOld(j).Xi FloeOld(j).Yi]);
        end
        floenew = [];
        for ii = 1:length(Floe)
            ParFloes(ii).floenew = [];
            ParFloes(ii).kill = [];
            ParFloes(ii).verts = 0;
        end
        if isfield(Floe,'poly')
            Floe=rmfield(Floe,{'poly'});
        end
        parfor ii = 1:length(Floe)
            floe = Floe(ii);
            if length(Floe(ii).c0) > 30
                [floe2,kill] = FloeSimplify(Floe(ii),0,FloeOld,polyboundary);
                if isfield(floe2,'poly')
                    floe2=rmfield(floe2,{'poly'});
                end
                if isempty(kill)
                    ParFloes(ii).kill = [ParFloes(ii).kill kill];
                end
                for jj = 1:length(floe2)
                    if jj == 1
                        Floe(ii) = floe2(jj);
                        ParFloes(ii).verts=length(floe2(jj).c0(1,:));
                    else
                        ParFloes(ii).floenew = [ParFloes(ii).floenew floe2(jj)];
                    end
                end
            end
        end
        clear FloeOld
        if isfield(Floe,'poly')
            Floe=rmfield(Floe,{'poly'});
        end
        floenew =[]; kill = [];
        for ii = 1:length(ParFloes)
            floenew = [floenew ParFloes(ii).floenew];
            kill = [kill ParFloes(ii).kill];
        end
        kill = unique(kill(kill>Nb));
        live = cat(1,Floe.alive);
        live(kill) = 0;
        Floe(live==0)=[]; 
        Floe =[Floe floenew];
    end

    
    if mod(i_step,nDTOut)==0  %plot the state after a number of timesteps
        

        [eulerian_data] = calc_eulerian_data(Floe,Nx,Ny,Nb,c2_boundary,PERIODIC);
        if ifPlot
            [fig] = plot_basic(fig, Time,Floe,ocean,c2_boundary_poly,Nb,PERIODIC);
%              saveas(fig,['./figs/' num2str(im_num,'%03.f') '.jpg'],'jpg');
        end
        
        %Average coarse eulerian data
        if AVERAGE
            SigXXa = SigXX/fix(nDTOut); SigYXa = SigYX/fix(nDTOut);
            SigXYa = SigXY/fix(nDTOut); SigYYa = SigYY/fix(nDTOut);
            Eux = Eux/fix(nDTOut); Evx = Evx/fix(nDTOut);
            Euy = Euy/fix(nDTOut); Evy = Evy/fix(nDTOut);
            U = U/fix(nDTOut); V = V/fix(nDTOut);
            dU = dU/fix(nDTOut); dV = dV/fix(nDTOut);
            Sig = Sig/fix(nDTOut); 
            mass = mass/fix(nDTOut);
        else
            SigXXa = squeeze(eulerian_data.stressxx); SigYXa = squeeze(eulerian_data.stressyx);
            SigXYa = squeeze(eulerian_data.stressxy); SigYYa = squeeze(eulerian_data.stressyy);
            Eux = squeeze(eulerian_data.strainux); Evx = squeeze(eulerian_data.strainvx);
            Euy = squeeze(eulerian_data.strainuy); Evy = squeeze(eulerian_data.strainvy);
            U = U+squeeze(eulerian_data.u);V = V+squeeze(eulerian_data.v);
            dU = dU+squeeze(eulerian_data.du);dV = dV+squeeze(eulerian_data.dv);
            Sig = Sig+squeeze(eulerian_data.stress);
        end
        
                
        if ifPlotStress
            [fig] =plot_basic_stress(fig, Time,Floe,ocean,c2_boundary_poly,Nb);
            saveas(fig,['./figs/' num2str(im_num,'%03.f') '.jpg'],'jpg');
            figure(2)
            plot(Xc,SigXXa,'kx','linewidth',2); title(['Time = ' num2str(Time/3600) ' hours'],'fontsize',24);
            drawnow
        end
        
        if ifPlotStressStrain
            fig2 = figure(fig2);
            SigO = Sig;
            subplot(2,4,1); imagesc(Xc,Yc,SigXXa); hold on; quiver(Xc,Yc,U*1e6,V*1e6,'k','autoscale','on'); set(gca,'YDir','normal','DataAspectRatio',[1 1 1]); title('$\sigma_{xx}$','interpreter','latex','fontsize',16); colorbar; caxis([-1e6 1e6])
            subplot(2,4,2); imagesc(Xc,Yc,SigYXa); hold on; quiver(Xc,Yc,U*1e6,V*1e6,'k','autoscale','on'); set(gca,'YDir','normal','DataAspectRatio',[1 1 1]); title('$\sigma_{yx}$','interpreter','latex','fontsize',16); colorbar; caxis([-1e6 1e6])
            subplot(2,4,5); imagesc(Xc,Yc,SigXYa); hold on; quiver(Xc,Yc,U*1e6,V*1e6,'k','autoscale','on'); set(gca,'YDir','normal','DataAspectRatio',[1 1 1]); title('$\sigma_{xy}$','interpreter','latex','fontsize',16); colorbar; caxis([-1e6 1e6])
            subplot(2,4,6); imagesc(Xc,Yc,SigYYa); hold on; quiver(Xc,Yc,U*1e6,V*1e6,'k','autoscale','on'); set(gca,'YDir','normal','DataAspectRatio',[1 1 1]); title('$\sigma_{yy}$','interpreter','latex','fontsize',16); colorbar; caxis([-1e6 1e6])
            subplot(2,4,3); imagesc(Xc,Yc,Eux); hold on; quiver(Xc,Yc,U*1e6,V*1e6,'k','autoscale','on'); set(gca,'YDir','normal','DataAspectRatio',[1 1 1]); title('$E_{11}$','interpreter','latex','fontsize',16); colorbar; caxis([-5e-6 5e-6])
            subplot(2,4,4); imagesc(Xc,Yc,Evx); hold on; quiver(Xc,Yc,U*1e6,V*1e6,'k','autoscale','on'); set(gca,'YDir','normal','DataAspectRatio',[1 1 1]); title('$E_{21}$','interpreter','latex','fontsize',16); colorbar; caxis([-5e-6 5e-6])
            subplot(2,4,7); imagesc(Xc,Yc,Euy); hold on; quiver(Xc,Yc,U*1e6,V*1e6,'k','autoscale','on'); set(gca,'YDir','normal','DataAspectRatio',[1 1 1]); title('$E_{12}$','interpreter','latex','fontsize',16); colorbar; caxis([-5e-6 5e-6])
            subplot(2,4,8); imagesc(Xc,Yc,Evy); hold on; quiver(Xc,Yc,U*1e6,V*1e6,'k','autoscale','on'); set(gca,'YDir','normal','DataAspectRatio',[1 1 1]); title('$E_{22}$','interpreter','latex','fontsize',16); colorbar; caxis([-5e-6 5e-6])
            saveas(fig2,['./figs/' num2str(im_num,'Stress%03.f') '.jpg'],'jpg');
        end
        
    end
    
    %Periodic creation of new elements in open ocean
    if PACKING && h0 > 0
        if mod(i_step,nDTpack)==0
            height.mean = h0;
            height.delta = 0;
            [Floe,Vd] = create_new_ice(Floe,c2_boundary,dhdt,Vd,target_concentration, height, min_floe_size, PERIODIC,3,3,Nb);
        end
    end
    
    %output and reset data values
    if mod(i_step,nDTOut)==0
        save(['./Floes/Floe' num2str(im_num,'%07.f') '.mat'],'Floe','eulerian_data','SigXXa','SigXYa', 'SigYYa','U','dU','mass');
        SigXX = zeros(Ny, Nx); SigYX = zeros(Ny, Nx);
        SigXY = zeros(Ny, Nx); SigYY = zeros(Ny, Nx);
        Eux = zeros(Ny, Nx); Evx = zeros(Ny, Nx);
        Euy = zeros(Ny, Nx); Evy = zeros(Ny, Nx);
        U = zeros(Ny, Nx); V = zeros(Ny, Nx);
        dU = zeros(Ny, Nx); dV = zeros(Ny, Nx);
        Sig = zeros(Ny, Nx); mass = zeros(Ny, Nx);
        
        M = cat(1,Floe.mass);
        Mtot(im_num) = sum(M)+sum(Vdnew(:));
        
        im_num=im_num+1;  %image number for saving data and coarse vars;
    end
    
    %Calculate forces and torques and intergrate forward
    [Floe,dissolvedNEW] = floe_interactions_all(Floe, floebound, ocean, winds, c2_boundary, dt, HFo,min_floe_size, Nx,Ny,Nb, dissolvedNEW,doInt,COLLISION, PERIODIC, RIDGING, RAFTING);

    %Average coarse eulerian data
    if AVERAGE
        [eulerian_data] = calc_eulerian_data(Floe,Nx,Ny,Nb,c2_boundary,PERIODIC);
        SigXX = SigXX+squeeze(eulerian_data.stressxx); SigYX = SigYX+squeeze(eulerian_data.stressyx);
        SigXY = SigXY+squeeze(eulerian_data.stressxy); SigYY = SigYY+squeeze(eulerian_data.stressyy);
        Eux = Eux+squeeze(eulerian_data.strainux); Evx = Evx+squeeze(eulerian_data.strainvx);
        Euy = Euy+squeeze(eulerian_data.strainuy); Evy = Evy+squeeze(eulerian_data.strainvy);
        U = U+squeeze(eulerian_data.u);V = V+squeeze(eulerian_data.v);
        dU = dU+squeeze(eulerian_data.du);dV = dV+squeeze(eulerian_data.dv); 
        Sig = Sig+squeeze(eulerian_data.stress);
        mass = mass+squeeze(eulerian_data.Mtot);
    end
    
    % Weld floes
    if WELDING && mod(i_step,25)==0 && dhdt > 0
        Fweld = 150;%Welding probability coefficient
        A=cat(1,Floe.area);
        if max(A) > Amax
           Amax = max(A); 
        end
        if mod(i_step,5000)==0
            Floe = weld(Floe,Nb,Fweld,c2_boundary,Amax/2,1,1);
        elseif mod(i_step,500)==0
            Floe = weld(Floe,Nb,Fweld,c2_boundary,Amax/3,2,2);
        else
            Floe = weld(Floe,Nb,Fweld,c2_boundary,Amax/3,3,3);
        end
    end
    
    % Fracture floes based upon stress
    if FRACTURES && mod(i_step,75)==0 
        compactness = 1; 
        [Floe,Princ] = fracture(Floe,Nb,min_floe_size,compactness);
    end
    
    % Remove sharp corners of interacting floes
    if CORNERS && mod(i_step,10)==0
        
        keep = rand(length(Floe),1)>0.7;
        keep(1:Nb) = ones(Nb,1); %Do not fracture boundaries
        if KEEP_MIN %Do not allow floes below threshold to fracture
            keep(cat(1,Floe.area)<min_floe_size)=1;
        end
        overlapArea=cat(1,Floe.OverlapArea)./cat(1,Floe.area);
        keep(overlapArea>0.15) = 0;
        fracturedFloes=corners(Floe(~keep),Nb,Floe,c2_boundary);
        if ~isempty(fracturedFloes)
            Floe=[Floe(keep) fracturedFloes];
        end
    end
    
    %Advect the dissolved mass
    Area=cat(1,Floe.area);
    if ~KEEP_MIN
        dissolvedNEW = calc_dissolved_mass(Floe(Area<min_floe_size),Nx,Ny,c2_boundary_poly)+dissolvedNEW;
    end
%     Vdnew = Advect_Dissolved_Ice(Vd,coarseMean,im_num,dissolvedNEW,c2_boundary,dt);
    Vdnew = Vd(:,:,1)+dissolvedNEW;
    dissolvedNEW=zeros(Ny,Nx);
    Vd(:,:,2) = Vd(:,:,1);
    Vd(:,:,1) = Vdnew;
    
    % Remove small floes from simulation
    Area=cat(1,Floe.area);
    if ~KEEP_MIN
        if sum(Area<min_floe_size)>0, display(['num of small floes killed:' num2str(sum(Area<min_floe_size))]); end
        Floe=Floe(Area> min_floe_size);
    end
    live = cat(1,Floe.alive);
    Floe(live == 0) = [];
    
    
    Time=Time+dt; i_step=i_step+1; %update time index

    
end

tEnd = toc(tStart)